### PR TITLE
Added support for PyErr_WriteUnraisable

### LINKED ISSUE #2884: Expose PyErr_WriteUnraisable
Since you sometimes might have some situations where Rust code that is infallible really likes to call into a Python callback, you end up with a horrible situation where all you can do is log the error.

However with Python 3.8 and later you can report such errors so they can be trapped. It would be nice to have a way to do this without unsafe code:

```rust
pub fn report_unraisable(py: Python<'_>, err: PyErr) {
    err.restore(py);
    unsafe {
        PyErr_WriteUnraisable(std::ptr::null_mut());
    }
}
```

I propose to add a `write_unraisable()` method directly onto `PyErr` for this purpose.

### PR DESCRIPTION
**Pull request by mitsuhiko**
Fixes #2884

### CODE REVIEW

**Review by davidhewitt — COMMENTED** [review]
Thanks! New API and test look fine to me, though I have some nits on the test code :)

Willing to fixup those and squash to a single commit? Then let's merge 👍

File: tests/test_exceptions.rs [hunk 1]
```diff
@@ -96,3 +97,50 @@ fn test_exception_nosegfault() {
...
+#[test]
+fn test_write_unraisable() {
+    Python::with_gil(|py| {
+        let report_unraisable = wrap_pyfunction!(report_unraisable)(py).unwrap();
+        let locals = PyDict::new(py);
+        locals
+            .set_item("report_unraisable", report_unraisable)
+            .unwrap();
+
+        let source = r#"if True:
```
[the remark this answers is not in the document]
Comment (mitsuhiko) [reply tests/test_exceptions.rs:123]: For some reason the `py_run!` macro gives me scope errors. I tried that originally.

[the remark this answers is not in the document]
Comment (davidhewitt) [reply tests/test_exceptions.rs:120]: Hmm, strange. I'll have a peek at that in the morning, I'd like to understand that.

File: src/err/mod.rs [hunk 2]
```diff
@@ -476,6 +476,15 @@ impl PyErr {
         unsafe { ffi::PyErr_Restore(ptype, pvalue, ptraceback) }
     }
 
+    /// Reports the error as unraisable.
+    ///
+    /// This calls `sys.unraisablehook()` using the current exception and obj argument.
```
Comment (mejrs) [inline src/err/mod.rs:481]: I'd like this documentation to be fleshed out a bit more. e.g.: What does an unraisable error do? when you you use it? when shouldn't you use it? An example would also be appreciated :)

**Comment by davidhewitt** [comment]
@mitsuhiko thanks again for this - I'll give it a quick rebase & fixup this morning so that hopefully it passes CI and merges. 👍

**Comment by davidhewitt** [comment]
I've rebased and rewritten the test to not use `py.run`.

**Pull request merged**